### PR TITLE
cql: don't crash when creating a view during a truncate

### LINKED ISSUE #17543: Assertion fails when truncating table with views being built
```
scylla: replica/table.cc:2330: future<db::replay_position> replica::table::discard_sstables(db_clock::time_point): Assertion `std::ranges::all_of(compaction_groups(), [this] (const compaction_group& cg) { return _compaction_manager.compaction_disabled(cg.as_table_state()); })' failed.
```

```
__pthread_kill_implementation at ??:?
__GI_raise at :?
__GI_abort at :?
__assert_fail_base.cold at ??:?
__assert_fail at ??:?
replica::table::discard_sstables(std::chrono::time_point<db_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >) at table.cc:?
replica::database::truncate(db::system_keyspace&, replica::table&, replica::database::table_truncate_state const&, std::chrono::time_point<db_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::$_0::operator()(view_ptr) const at database.cc:?
replica::database::truncate(db::system_keyspace&, replica::table&, replica::database::table_truncate_state const&, std::chrono::time_point<db_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >) [clone .resume] at database.cc:?
seastar::internal::coroutine_traits_base<void>::promise_type::run_and_dispose() at main.cc:?
seastar::reactor::run_some_tasks() at ??:?
seastar::reactor::do_run() at ??:?
seastar::reactor::run() at ??:?
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) at ??:?
seastar::app_template::run(int, char**, std::function<seastar::future<int> ()>&&) at ??:?
scylla_main(int, char**) at main.cc:?
std::function<int (int, char**)>::operator()(int, char**) const at ??:?
main at main.cc:?
__libc_start_call_main at ??:?
__libc_start_main_alias_2 at :?
_start at ??:?
```

Found with extension to dtest (PR will come soon)

### PR DESCRIPTION
**Pull request by nyh**
The test dtest materialized_views_test.py::TestMaterializedViews:: test_mv_populating_from_existing_data_during_truncate reproduces an assertion failure, and crash, while doing a CREATE MATERIALIZED VIEW during a TRUNCATE operation.

This patch fixes the crash by removing the assert() call - we'll explain below why this is fine. It makes the test stop crashing Scylla, but it still fails due to issue #17293, so that issue must be fixed before this patch is merged (because as soon as this patch is merged, #17543 will be marked fixed, the test will start running in CI).

Let's explain the crash, and the fix:

The test starts TRUNCATE on table that doesn't yet have a view. truncate_table_on_all_shards() begins by disabling compaction on the table and all its views (of which there are none, at this point). At this point, the test creates a new view is on this table. The new view has, by default, compaction enabled. Later, TRUNCATE calls discard_sstables() on this new view, asserts that it has compaction disabled - and this assertion fails.

The fix in this patch is to remove the assert(). In other words, we acknowledge that in this use case, the view *will* have compactions enabled while being truncated. I claim that this is "good enough", if we remember *why* we disable compaction in the first place: It's important to disable compaction while truncating because truncating during compaction can lead us to data resurection when the old sstable is deleted during truncation but the result of the compaction is written back. True, this can now happen in a new view (a view created *DURING* the truncation). But I claim that worse things can happen for this new view: Notably, we may truncate a view and then the ongoing view building (which happens in a new view) might copy data from the base to the view and only then truncate the base - ending up with an empty base and non-empty view. This problem is more likely, and more serious, than the compaction problem, so will need to be solved in a separate patch.

Fixes #17543.

### CODE REVIEW

**Review by xemul — APPROVED** [review]
This does makes test fail later and in a more gentle way -- by seeing unexpected rows in view table, but that's another story

**Comment by nyh** [comment]
Please don't merge this PR yet - it will cause next promotion failure. this PR shouldn't have passed dtest, because once it is merged a `@requires` dtest will start running and fail.
(by the way, @fruch, this is a minor bug in how `@requires` works, it can cause CI to pass but once the PR is **really** merged, the test will start failing.)

Before merging this PR we need to merge https://github.com/scylladb/scylla-dtest/pull/4029 to dtest - to make the test `@require` another issue, so it won't start running (and failing) when this PR is merged and closes its issue.

**Comment by fruch** [comment]
> Please don't merge this PR yet - it will cause next promotion failure. this PR shouldn't have passed dtest, because once it is merged a `@requires` dtest will start running and fail.
> (by the way, @fruch, this is a minor bug in how `@requires` works, it can cause CI to pass but once the PR is **really** merged, the test will start failing.)

This feature was planned for a daily run, and not for gating

I would argue that once a test is marked as such, it shouldn't be gating at all.

Also we don't have any way of knowing something hit next

> 
> Before merging this PR we need to merge https://github.com/scylladb/scylla-dtest/pull/4029 to dtest - to make the test `@require` another issue, so it won't start running (and failing) when this PR is merged and closes its issue.

**Comment by nyh** [comment]
> > Please don't merge this PR yet - it will cause next promotion failure. this PR shouldn't have passed dtest, because once it is merged a `@requires` dtest will start running and fail.
> > (by the way, @fruch, this is a minor bug in how `@requires` works, it can cause CI to pass but once the PR is **really** merged, the test will start failing.)
> 
> This feature was planned for a daily run, and not for gating

Oh, right, I see. I just assumed without checking that the test we're discussing and has `@requires` - materialized_views_test.py::TestMaterializedViews::test_mv_populating_from_existing_data_during_truncate - was in gating. Is it? I can't figure out.
 
> I would argue that once a test is marked as such, it shouldn't be gating at all.

I agree.

**Comment by fruch** [comment]
> > > Please don't merge this PR yet - it will cause next promotion failure. this PR shouldn't have passed dtest, because once it is merged a `@requires` dtest will start running and fail.
> > > (by the way, @fruch, this is a minor bug in how `@requires` works, it can cause CI to pass but once the PR is **really** merged, the test will start failing.)
> > 
> > 
> > This feature was planned for a daily run, and not for gating
> 
> Oh, right, I see. I just assumed without checking that the test we're discussing and has `@requires` - materialized_views_test.py::TestMaterializedViews::test_mv_populating_from_existing_data_during_truncate - was in gating. Is it? I can't figure out.

the whole module is taken out of gating:
https://github.com/scylladb/scylla-dtest/commit/17f9d015868ffb6fb27595099ce63dcc502832c7

so in this case it's currently not

> 
> > I would argue that once a test is marked as such, it shouldn't be gating at all.
> 
> I agree.

here a PR for doing just that, i.e. skipping test marked with require when running as gating:
https://github.com/scylladb/scylla-dtest/pull/4031

**Comment by nyh** [comment]
> the whole module is taken out of gating: [scylladb/scylla-dtest@17f9d01](https://github.com/scylladb/scylla-dtest/commit/17f9d015868ffb6fb27595099ce63dcc502832c7)

The entire collection of materialized views tests were taken out of gating? Isn't this excessive?

**Comment by fruch** [comment]
> > the whole module is taken out of gating: [scylladb/scylla-dtest@17f9d01](https://github.com/scylladb/scylla-dtest/commit/17f9d015868ffb6fb27595099ce63dcc502832c7)
> 
> The entire collection of materialized views tests were taken out of gating? Isn't this excessive?

https://github.com/scylladb/scylladb/issues/15046#issuecomment-1712782923

the request was to take it, and that someone (i.e. whom owns this feature), would continue the investigation / fixing, no ontop of gating.

as that thread keeps ongoing, but I don't think anyone did any stabilize those back into gating. (or even part of them)

**Comment by nyh** [comment]
I merged the dtest fix (https://github.com/scylladb/scylla-dtest/pull/4029) so as far as I'm concerned, this PR can go in.

File: replica/table.cc
Comment (avikivity) [inline replica/table.cc:2335]: Shouldn't we at least warn? Or check if it's a view?

Comment (nyh) [reply replica/table.cc:2335]: I can change the assert to on_internal_error, and leave it if the table is not a view. It still feels very wrong to me, but I guess it may catch some regression in the future that has nothing to do with views. Ok.

Comment (mykaul) [reply replica/table.cc:2335]: I'm not sure I see the value - what will happen when we bump into on_internal_error()? It'll crash on QA and in production throw an exception?

Comment (avikivity) [reply replica/table.cc:2335]: Yes

**Comment by nyh** [comment]
@dani-tweig @mykaul my PR just failed on a "PR require backport label / label" :-( Can you guys **please** decided if the label needs to go on the PR (according to @dani-tweig) or the issue (according to @mykaul) before making this mandatory?

File: replica/table.cc
Comment (nyh) [reply replica/table.cc]: Pushed a new version with on_internal_error() if not a view, ignored when a view.

If on_internal_error() fails in production, the user will see a failure (exception) in the "TRUNCATE" operation, which, arguably, exactly what we expect to happen. I'm not sure, though, what happens if we have this failure on just one node and not others - I think we can end up with a half-truncated state, but don't think there's a better solution - a crash is worse.

**Comment by nyh** [comment]
@benipeled @dani-tweig I think there's another process bug here: Because I didn't put a "backport" tag on this PR, it was rejected without building. Then, when I added the backport tag, it checked the tag and marked success in this check - but did not proceed with the build! So now this PR is missing CI results :-(

**Comment by dani-tweig** [comment]
@yaronkaikov 
Can you have a look?

**Comment by yaronkaikov** [comment]
> @yaronkaikov Can you have a look?



> @benipeled @dani-tweig I think there's another process bug here: Because I didn't put a "backport" tag on this PR, it was rejected without building. Then, when I added the backport tag, it checked the tag and marked success in this check - but did not proceed with the build! So now this PR is missing CI results :-(

What do you mean? The backport label is not related to CI in any way. It shouldn't hold CI from starting the build + testing process

**Comment by yaronkaikov** [comment]
> > @yaronkaikov Can you have a look?
> 
> > @benipeled @dani-tweig I think there's another process bug here: Because I didn't put a "backport" tag on this PR, it was rejected without building. Then, when I added the backport tag, it checked the tag and marked success in this check - but did not proceed with the build! So now this PR is missing CI results :-(
> 
> What do you mean? The backport label is not related to CI in any way. It shouldn't hold CI from starting the build + testing process

Or did you mean this
image

**Comment by nyh** [comment]
> @benipeled @dani-tweig I think there's another process bug here: Because I didn't put a "backport" tag on this PR, it was rejected without building. Then, when I added the backport tag, it checked the tag and marked success in this check - but did not proceed with the build! So now this PR is missing CI results :-(

It seems eventually, all the CI ran. When I wrote this comment, only the "PR require backport label" check was done in the green checkbox in the bottom - and none of the other lines appeared. I waited and waited, and they didn't appear. I guess I should have waited longer - the tests just took a really long time to start?

I said this in the past, and want to suggest this again: all the tests that we're going to run should be listed as "pending" as soon as possible, if we can't do this in seconds (that would have been best...) then please in a couple of minutes - if I don't see them in 10 minutes, I think something is wrong. It's ok (but not great...) that the tests still need to wait an hour to get a VM or something,but we should add the lines to the UI even before we get those final VMs.

**Comment by nyh** [comment]
@scylladb/scylla-maint please consider merging. @avikivity I already pushed a new version based on your review - the new version does an on_internal_error() if not a view, so it doesn't lose the old sanity-check for the usual case of base tables.

**Comment by nyh** [comment]
Push a new version - rebased and the typo found by @denesb fixed.
Please consider merging.

**Review by denesb — APPROVED** [review]

**Comment by mykaul** [comment]
@yaronkaikov - how can we ask nicely the backport bot to run on this?

**Comment by yaronkaikov** [comment]
Bot already did the work
5.4 - https://github.com/scylladb/scylladb/pull/17928
5.2 - https://github.com/scylladb/scylladb/pull/17929

**Comment by mykaul** [comment]
> Bot already did the work 5.4 - #17928 5.2 - #17929

@yaronkaikov - both have conflicts, who owns to resolve them and push them forward?

**Comment by yaronkaikov** [comment]
> > Bot already did the work 5.4 - #17928 5.2 - #17929
> 
> @yaronkaikov - both have conflicts, who owns to resolve them and push them forward?

Both were automatically assigned to PR original author which is @nyh

**Comment by nyh** [comment]
New PR for backport to 5.4: https://github.com/scylladb/scylladb/pull/19580

**Pull request closed without merging**
